deprecation/__init__.pyi

```diff
@@ -0,0 +1,33 @@
+"""Stubs file for `deprecation.py`.
+
+Created using mypy's `stubgen`, types manually corrected.
+"""
+
+import datetime
+from packaging import version
+from typing import Callable, Optional, Union
+
+message_location: str
+
+class DeprecatedWarning(DeprecationWarning):
+    function: str
+    deprecated_in: Union[datetime.date, version.Version, None]
+    removed_in: Union[datetime.date, version.Version, None]
+    details: Union[datetime.date, version.Version, None]
+    def __init__(
+        self,
+        function: str,
+        deprecated_in: Union[datetime.date, version.Version, None],
+        removed_in: Union[datetime.date, version.Version, None],
+        details: str = ...,
+    ) -> None: ...
+
+class UnsupportedWarning(DeprecatedWarning): ...
+
+def deprecated(
+    deprecated_in: Union[str, datetime.date, version.Version, None] = ...,
+    removed_in: Union[str, datetime.date, version.Version, None] = ...,
+    current_version: Union[str, datetime.date, version.Version, None] = ...,
+    details: str = ...,
+) -> Callable: ...
```

Comment:
Consider using a TypeVar for the wrapped function so that the decorator's type hints preserve the function's signature. Pylance (in VS Code) manages to display the signature without this (perhaps due to the implementation's use of `@functools.wraps`?), but mypy shows `Any` if I use `reveal_type` on a deprecated function.

Example: https://github.com/python/typeshed/blob/main/stdlib/contextlib.pyi#L34

```python
_F = TypeVar("_F", bound=Callable[..., Any])

def deprecated(...) -> Callable[[_F], _F]: ...
def fail_if_not_removed(method: _F) -> _F: ...
```